[WIP] Add code for training turn taking prediction model

siddhu001 opened the pull request:
## What?
1. Implemented code for training and inference of the encoder + classifier model.
2. Added data preparation and evaluation for training the turn-taking model using the Switchboard dataset.
3. Developed functionality for turn-taking evaluation metrics.

## Why?
1. Enabled the training and evaluation of the turn-taking model on the Switchboard dataset.
2. Added support for using the trained turn-taking model to assess human-AI conversations.

Note: This PR is a work in progress, and code cleanup is ongoing.

suckor commented:
hi, guys, I noticed that the channel information is ignored in the local/create_espnet_data_folders.py file, and the generated wav.scp only uses mono-data. Is there a problem?

siddhu001 commented:
Hi,

Thank you so much for your interest in our work and for reaching out!

This PR is currently a work in progress, so it’s not yet fully documented. However, here’s an overview of the current workflow for running the turn-taking prediction model:

Data Preparation Script:
The data preparation bash script is located at local/data_turn_take.sh.

1. Stages 1 & 2: These stages are similar to the Switchboard data preparation process.
2. Stage 3:

- a. The script first calls local/create_switchboard_data_2channels.py, which annotates each 40 ms chunk with turn-taking labels specific to each speaker.
- b. Next, it calls local/create_switchboard_data_2channels_mono.py, which generates speaker-invariant turn-taking labels for each 40 ms chunk. Labeling scheme: NA: No one is talking. BC (Backchannel): One speaker backchannels during the other speaker’s turn. I (Interruption): Both speakers are talking simultaneously to take the turn, without backchanneling. T (Turn Change): There is a turn change. C (Continuation): Any other case.
- c. ESPnet Data Folder Creation: The script then calls local/create_espnet_data_folders.py, which generates single-channel "mixed" audio as input for the turn-taking model.

Our turn-taking model is designed to take single-channel mixed speech as input and predict speaker-invariant turn-taking labels. We decided on this design choice since separate channel speech for each speaker, which might not always be available. By effectively modeling both speakers’ turn-taking behavior from single-channel mixed speech, our approach overcomes this limitation.

I plan to prioritize documenting this PR to enhance its readability and usability in the coming weeks. Once the documentation is complete, I aim to merge it into the ESPnet codebase. If you could check back in a few weeks, the codebase should be much easier to follow.

I sincerely apologize for any inconvenience this may cause and truly appreciate your patience and interest in our work. Please don’t hesitate to reach out if you have any further questions or need clarification.

suckor commented:
> Hi,
> 
> Thank you so much for your interest in our work and for reaching out!
> 
> This PR is currently a work in progress, so it’s not yet fully documented. However, here’s an overview of the current workflow for running the turn-taking prediction model:
> 
> Data Preparation Script: The data preparation bash script is located at local/data_turn_take.sh.
> 
> 1. Stages 1 & 2: These stages are similar to the Switchboard data preparation process.
> 2. Stage 3:
> 
> * a. The script first calls local/create_switchboard_data_2channels.py, which annotates each 40 ms chunk with turn-taking labels specific to each speaker.
> * b. Next, it calls local/create_switchboard_data_2channels_mono.py, which generates speaker-invariant turn-taking labels for each 40 ms chunk. Labeling scheme: NA: No one is talking. BC (Backchannel): One speaker backchannels during the other speaker’s turn. I (Interruption): Both speakers are talking simultaneously to take the turn, without backchanneling. T (Turn Change): There is a turn change. C (Continuation): Any other case.
> * c. ESPnet Data Folder Creation: The script then calls local/create_espnet_data_folders.py, which generates single-channel "mixed" audio as input for the turn-taking model.
> 
> Our turn-taking model is designed to take single-channel mixed speech as input and predict speaker-invariant turn-taking labels. We decided on this design choice since separate channel speech for each speaker, which might not always be available. By effectively modeling both speakers’ turn-taking behavior from single-channel mixed speech, our approach overcomes this limitation.
> 
> I plan to prioritize documenting this PR to enhance its readability and usability in the coming weeks. Once the documentation is complete, I aim to merge it into the ESPnet codebase. If you could check back in a few weeks, the codebase should be much easier to follow.
> 
> I sincerely apologize for any inconvenience this may cause and truly appreciate your patience and interest in our work. Please don’t hesitate to reach out if you have any further questions or need clarification.

Thanks for your reply! I didn't notice it used mixed speech. 
This is a very interesting work!  I look forward to more updates in the future.

siddhu001 commented:
Hi @sw005320,

I have cleaned the codebase, added documentation, and integrated unit tests. Specifically, this PR provides:

1. The training pipeline for our Judge Turn-Taking Model.
2. The inference code to generate turn-taking pseudo-labels using our model.
3. The evaluation platform to compute our proposed turn-taking metrics.

I would really appreciate it if you could review the PR and merge it if everything looks good. Thanks a lot!

sw005320 commented:
Hi @siddhu001,

I'm a bit hesitant to add this to the main `asr.sh` as it is more than ASR.
Is it possible to implement it in `slu.sh`?

siddhu001 commented:
> Hi @siddhu001,
> 
> I'm a bit hesitant to add this to the main `asr.sh` as it is more than ASR. Is it possible to implement it in `slu.sh`?

I see your point @sw005320 , and I agree that this maybe aligns more with Spoken Language Understanding. I'll start working on porting it into espnet2/slu. Thanks for your feedback!

siddhu001 commented:
> > Hi @siddhu001,
> > I'm a bit hesitant to add this to the main `asr.sh` as it is more than ASR. Is it possible to implement it in `slu.sh`?
> 
> I see your point @sw005320 , and I agree that this maybe aligns more with Spoken Language Understanding. I'll start working on porting it into espnet2/slu. Thanks for your feedback!

@sw005320 I’ve completed porting this to the SLU codebase. Could you kindly review the PR again and merge it if everything looks good? Please let me know if any further changes are needed. Thanks a lot!

sw005320 commented on espnet2/bin/slu_inference.py:
```diff
@@ -272,6 +273,58 @@ def __call__(
+
+        if self.run_chunk:
+            speech = to_device(speech, device=self.device)
+            sim_chunk_length = 640
+            start_chunk = 3200
```
It's better to make them controlled by arguments.

siddhu001 replied:
Done!

sw005320 commented:
@jctian98, can I ask you to review this PR?

siddhu001 commented:
@jctian can you please review this PR asap? This is needed for one of speech technology for conversation AI course project.

sw005320 commented:
@jctian98, this is a reminder.

jctian98 commented on egs2/swbd/slu1/local/turn_take_utils.py:
```diff
@@ -0,0 +1,237 @@
+                if word_file_A[count_A][0] != sentence_dict_sort[k][2]:
+                    import pdb
+
+                    pdb.set_trace()
```
redundent

siddhu001 replied:
Thanks for pointing this out! I fixed this now!

jctian98 commented on egs2/swbd/slu1/local/score.sh:
Is this file necessary? can we just call the python file?
It's ok if just for compatibility.

siddhu001 replied:
This is required for compatibility, as slu.sh calls score.sh. The score.sh script is dataset-specific, and each data recipe must include its own version of score.sh with the appropriate SLU evaluation metrics.

jctian98 commented:
@siddhu001 I may not have sufficient expertise to understand the details of turn-taking. But in general, I think the PR is good. It contains documents and test cases.

Just one thing: it seems many pdb-related test code is not cleaned up. Maybe we can fix it.

siddhu001 commented:
> @siddhu001 I may not have sufficient expertise to understand the details of turn-taking. But in general, I think the PR is good. It contains documents and test cases.

Thanks so much for taking the time to review the PR!

> Just one thing: it seems many pdb-related test code is not cleaned up. Maybe we can fix it.

You're absolutely right — the pdb statements were initially added to help flag unexpected behavior during debugging. Based on your suggestion, I’ve now replaced all instances of pdb with sys.exit(1) to more clearly indicate failure cases.

As for the CI failures, they seem to be related to an unrelated ESPnet installation issue. From my end, I believe I’ve now addressed all the review comments. Please let me know if there's anything else you'd like me to update!

siddhu001 commented:
@sw005320 can this PR please be merged? All CI have passed and I have incorporated all the feedback.

sw005320 commented:
Sure. Thanks for your contributions!
